compiler/rustc_query_system/src/dep_graph/graph.rs

```diff
@@ -115,9 +125,68 @@ impl<K: DepKind> DepGraph<K> {
         encoder: FileEncoder,
         record_graph: bool,
         record_stats: bool,
+        shared: bool,
     ) -> DepGraph<K> {
         let prev_graph_node_count = prev_graph.node_count();
 
+        let colors = DepNodeColorMap::new(prev_graph_node_count);
+
+        if !shared {
+            let current = DepGraph::get_current::<SRefCell>(
```

Comment:
The codegen across both SRefCell and SMutex of DepGraph and other types(?) already seems to increase the compile-times (and probably binary size) with this PR, and I'd guess presumably that will continue with this approach...

Reply:
I think we can concentrate these code spanning `SMutex` and `SRefCell` where they are needed most, such as querying DepNodes in the system and caching data structures. to minimize the impact on binary size. Maybe we can use some data structure design to avoid this to some extent, like `WorkerLocal`.